modules/primo/src/Primo/Ting/Result.php

```diff
@@ -133,8 +143,35 @@ public function getSearchRequest() {
    *   List of facets, empty if none where found.
    */
   public function getFacets() {
-    // TODO: Implement.
-    return [];
+    // Return the mapped facets if we've already done the work.
+    if ($this->facets !== NULL) {
+      return $this->facets;
+    }
+
+    // Prepare for loading.
+    $this->facets = [];
+
+    // Handle empty facets.
+    $primo_facets = $this->result->getFacets();
+    if (empty($primo_facets)) {
+      return $this->facets;
+    }
+
+    // Convert Primo Facets to TingSearchFacetTerm instances.
+    array_walk($primo_facets, function (Facet $facet) {
+      $terms = [];
+      $values = $facet->getValues();
+      array_walk($values, function ($frequency, $name) use (&$terms) {
+        $terms[] = new TingSearchFacetTerm($name, $frequency);
+      });
+
+      // Return facets indexed by their id prefixed with "facet_". This will
+      // make the id match the syntax Primo expects when it is later used as a
+      // field query in ting_search_search_execute().
+      $this->facets['facet_' . $facet->getId()] = new TingSearchFacet($facet->getId(), $terms);
```

Comment:
:+1: for the explanation.

Would it make sense to set the id for the `TingSearchFacet` to `facet_' . $facet->getId()` as well?